compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,185 +75,505 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
-  rewriter.setInsertionPoint(npuDmaOp);
-  if (scf::ForOp loop = npuDmaOp->getParentOfType<scf::ForOp>();
-      loop && getNumberIterations(loop)) {
-    // In a loop, using the semi-affine expression:
-    // `iv % size + offset`,
-    // `iv` is the loop induction variable,
-    // `size` is the number of BD IDs assigned to each DMA op,
-    // `offset` is the BD ID assigned to current DMA op at first iteration.
-    Value iv = loop.getInductionVar();
-
-    // Get the number of BD IDs will be assigned to current DMA op.
-    uint32_t numRequired = 0;
-    getNumRequiredBdIds(loop, npuDmaOp, tileOp, shimTileToGeneratorMap,
-                        numRequired);
-    uint32_t numAvailable = generator.getNumAvailableBdIds(channel);
-    uint32_t size = std::max(numAvailable / numRequired, 1u);
-
-    // Only create expression if more than 1 BD ID is needed,
-    // otherwise, fall back to constant BD ID.
-    if (size > 1) {
-      // Assigning BD IDs for all iterations in the loop.
-      SmallVector<uint32_t> bdIds;
-      for (uint32_t i = 0; i < size; i++) {
+
+  /// Given a DmaBatch `outerDmaBatch`, traverse each nested DmaBatch in it and
+  /// infer the required Bd Ids for them. If the nested DmaBatch is surrounded
+  /// with a scf.for, we account for that while sending the required Bd Ids to
+  /// the caller of this API.
+  int32_t getRequiredBdIdsForInnerBatches() {
+    if (immediateInnerBatches.empty()) return 0;
+    int32_t requiredBdIds = 0;
+    for (const std::unique_ptr<DmaBatch> &dmaBatch : immediateInnerBatches) {
+      DmaBatch *currDmaBatch = dmaBatch.get();
+      do {
+        if (currDmaBatch->isEmpty()) break;
+        int32_t totalDmaOpsInCurrentBatch = currDmaBatch->currentDmaOps.size();
+        int32_t requiredBdIdsForInnerBatches =
+            currDmaBatch->getRequiredBdIdsForInnerBatches();
+        currDmaBatch->requiredBdIds =
+            requiredBdIdsForInnerBatches + totalDmaOpsInCurrentBatch;
+        requiredBdIds += totalDmaOpsInCurrentBatch;
+        if (currDmaBatch->forOpParent) {
+          std::optional<uint32_t> numIterations =
+              getNumberIterations(currDmaBatch->forOpParent);
+          if (numIterations) {
+            requiredBdIds +=
+                requiredBdIdsForInnerBatches * numIterations.value();
+          } else {
+            requiredBdIds += requiredBdIdsForInnerBatches;
+          }
+        }
+        currDmaBatch = currDmaBatch->nextDmaBatch.get();
+      } while (currDmaBatch != nullptr);
+    }
+    return requiredBdIds;
+  }
+};
+
+struct TileDmaBatchGraph {
+ private:
+  /// The main tile->DmaBatch graph.
+  llvm::MapVector<AMDAIE::TileOp, std::unique_ptr<DmaBatch>> tileDmaBatchGraph;
+  /// Helps keep track of the current DmaBatch of a tile that's being processed.
+  /// It holds the reference to the last DmaBatch for every tile on the main
+  /// graph.
+  DenseMap<AMDAIE::TileOp, DmaBatch *> currentTileBatch;
```

Comment:
If you really need multiple references to the same `DmaBatch`, you should use a `shared_ptr` instead.